### ISSUE TITLE
Monadic annotation prevents primitive step

### ISSUE THREAD

**Issue by cpitclaudel** [FStarLang/FStar#1305]
Is this an unfortunate side-effect of a previous fix?

```
let _ =
  assert_by_tactic True
                   (fun () ->
                     admit ();
                     let x = quote 1 () in
                     print (term_to_string (norm_term [] x ())) ();
                     fail "" ())
```

```
Tactic got stuck! Please file a bug report with a minimal reproduction of this issue.
(match
    match
      reify ((((fun _ ->
                  FStar.Tactics.Effect.TAC?.reflect (FStar.Tactics.Builtins.__print
                      (FStar.Reflection.Basic.__term_to_string
                        (FStar.Tactics.Builtins.__embed (PatternMatching.hyp Prims.l_True)))))
              <:
              uu___731931:Prims.unit -> FStar.Tactics.Effect.Tac Prims.unit)
            <:
            Prims.PURE (uu___731933:Prims.unit -> FStar.Tactics.Effect.Tac Prims.unit)) ())
        "(alien:tactics.embed_proofstate)"
    with
    | FStar.Tactics.Result.Success (_, q) ->
      (let () = FStar.Tactics.Types.tracepoint q in
        reify (FStar.Tactics.Derived.fail "" ()) (FStar.Tactics.Types.decr_depth q))
      <:
      FStar.Tactics.Result.__result Prims.unit
    | FStar.Tactics.Result.Failed (msg, q) -> FStar.Tactics.Result.Failed (msg, q)
  with
  | FStar.Tactics.Result.Success (a, q) ->
    (let () = FStar.Tactics.Types.tracepoint q in
      reify (FStar.Tactics.Effect.return a ()) (FStar.Tactics.Types.decr_depth q))
    <:
    FStar.Tactics.Result.__result Prims.unit
  | FStar.Tactics.Result.Failed (msg, q) -> FStar.Tactics.Result.Failed (msg, q))
<:
FStar.Tactics.Result.__result Prims.unit

```

**Comment by cpitclaudel** [comment]
Another example:

```
Tactic got stuck! Please file a bug report with a minimal reproduction of this issue.
(match
    match
      match
        match
          match
            FStar.Tactics.Builtins.__print
              ("Got binder: " ^ ("a#2431267" <: Prims.PURE Prims.string) ^ "; type is Type")
              "(alien:tactics.embed_proofstate)"
          with
          | FStar.Tactics.Result.Success (_, q) ->
            (let () = FStar.Tactics.Types.tracepoint q in
              FStar.Tactics.Effect._dm4f_TAC_return_elab
                (FStar.Reflection.Types.binder * Prims.string * FStar.Reflection.Types.term)
                ((let binder_kind, typ =
                      PatternMatching.classify_abspat_binder "(alien:reflection.embed_binder)"
                    in
                    "(alien:reflection.embed_binder)",
                    (FStar.Reflection.Basic.inspect_bv "(alien:reflection.embed_binder)"
                      <:
                      Prims.PURE Prims.string),
                    binder_kind,
                    typ)
                  <:
                  FStar.Reflection.Types.binder * Prims.string * FStar.Reflection.Types.term)
                (FStar.Tactics.Types.decr_depth q))
            <:
            FStar.Tactics.Result.__result
            (FStar.Reflection.Types.binder * Prims.string * FStar.Reflection.Types.term)
          | FStar.Tactics.Result.Failed (msg, q) -> FStar.Tactics.Result.Failed (msg, q)
        with
        | FStar.Tactics.Result.Success (a, q) ->
          (let () = FStar.Tactics.Types.tracepoint q in
            FStar.Tactics.Effect._dm4f_TAC_bind_elab
              (Prims.list (FStar.Reflection.Types.binder * Prims.string *
                    FStar.Reflection.Types.term))
              (Prims.list (FStar.Reflection.Types.binder * Prims.string *
                    FStar.Reflection.Types.term))
              _
              (reify (PatternMatching.Utils.tacmap (fun binder ->
                          ((let uu___2431431 =
                                PatternMatching.debug ("Got binder: " ^
                                    (FStar.Reflection.Basic.inspect_bv binder
                                      <:
                                      Prims.PURE Prims.string) ^
                                    "; type is " ^
                                    FStar.Reflection.Basic.term_to_string (FStar.Reflection.Basic.type_of_binder 
                                          binder))
                                <:
                                FStar.Tactics.Effect.TAC Prims.unit
                              in
                              ((let binder_kind, typ = PatternMatching.classify_abspat_binder binder
                                  in
                                  binder,
                                  (FStar.Reflection.Basic.inspect_bv binder
                                    <:
                                    Prims.PURE Prims.string),
                                  binder_kind,
                                  typ)
                                <:
                                FStar.Reflection.Types.binder * Prims.string *
                                FStar.Reflection.Types.term)
                              <:
                              Prims.PURE
                              (FStar.Reflection.Types.binder * Prims.string *
                                FStar.Reflection.Types.term))
                            <:
                            FStar.Tactics.Effect.TAC
                            (FStar.Reflection.Types.binder * Prims.string *
                              FStar.Reflection.Types.term))
                          <:
                          FStar.Tactics.Effect.TAC
                          (FStar.Reflection.Types.binder * Prims.string *
                            FStar.Reflection.Types.term))
                      ["(alien:reflection.embed_binder)"]))
              _
              (fun uu___2431434 ->
                  FStar.Tactics.Effect._dm4f_TAC_return_elab
                    (Prims.list (FStar.Reflection.Types.binder * Prims.string *
                          FStar.Reflection.Types.term)) (a :: uu___2431434))
              (FStar.Tactics.Types.decr_depth q))
          <:
          FStar.Tactics.Result.__result
          (Prims.list (FStar.Reflection.Types.binder * Prims.string * FStar.Reflection.Types.term))
        | FStar.Tactics.Result.Failed (msg, q) -> FStar.Tactics.Result.Failed (msg, q)
      with
      | FStar.Tactics.Result.Success (a, q) ->
        (let () = FStar.Tactics.Types.tracepoint q in
          FStar.Tactics.Effect._dm4f_TAC_bind_elab PatternMatching.matching_problem
            (PatternMatching.matching_problem * PatternMatching.abspat_continuation)
            _
            (reify (PatternMatching.Utils.tacfold_left (fun problem uu___2431440 ->
                        ((let binder, bv_name, binder_kind, typ = uu___2431440 in
                            (PatternMatching.debug ("Compiling binder " ^
                                  FStar.Reflection.Basic.inspect_bv binder ^
                                  ", classified as " ^
                                  PatternMatching.string_of_abspat_binder_kind binder_kind ^
                                  ", with type " ^ FStar.Reflection.Basic.term_to_string typ)
                              <:
                              FStar.Tactics.Effect.TAC Prims.unit;
                              ((match binder_kind with
                                  | PatternMatching.ABKVar _ ->
                                    PatternMatching.Mkmatching_problem
                                      (bv_name :: Mkmatching_problem?.mp_vars problem)
                                      (Mkmatching_problem?.mp_hyps problem)
                                      (Mkmatching_problem?.mp_goal problem)
                                    <:
                                    Prims.PURE PatternMatching.matching_problem
                                  | PatternMatching.ABKHyp ->
                                    ((let uu___2431447 =
                                          (let uu___2431448 =
                                              (let uu___2431449 =
                                                  PatternMatching.pattern_of_term typ
                                                  <:
                                                  FStar.Tactics.Effect.TAC
                                                  PatternMatching.Types.pattern
                                                in
                                                (bv_name, uu___2431449
                                                  <:
                                                  Prims.PURE
                                                  (PatternMatching.Types.varname *
                                                    PatternMatching.Types.pattern))
                                                <:
                                                FStar.Tactics.Effect.TAC
                                                (PatternMatching.Types.varname *
                                                  PatternMatching.Types.pattern))
                                              <:
                                              FStar.Tactics.Effect.TAC
                                              (PatternMatching.Types.varname *
                                                PatternMatching.Types.pattern)
                                            in
                                            (uu___2431448 :: Mkmatching_problem?.mp_hyps problem
                                              <:
                                              Prims.PURE
                                              (Prims.list (PatternMatching.Types.varname *
                                                    PatternMatching.Types.pattern)))
                                            <:
                                            FStar.Tactics.Effect.TAC
                                            (Prims.list (PatternMatching.Types.varname *
                                                  PatternMatching.Types.pattern)))
                                          <:
                                          FStar.Tactics.Effect.TAC
                                          (Prims.list (PatternMatching.Types.varname *
                                                PatternMatching.Types.pattern))
                                        in
                                        (PatternMatching.Mkmatching_problem
                                            (Mkmatching_problem?.mp_vars (problem
                                                  <:
                                                  Prims.PURE PatternMatching.matching_problem))
                                            uu___2431447
                                            (Mkmatching_problem?.mp_goal (problem
                                                  <:
                                                  Prims.PURE PatternMatching.matching_problem))
                                          <:
                                          Prims.PURE PatternMatching.matching_problem)
                                        <:
                                        FStar.Tactics.Effect.TAC PatternMatching.matching_problem)
                                      <:
                                      FStar.Tactics.Effect.TAC PatternMatching.matching_problem)
                                    <:
                                    FStar.Tactics.Effect.TAC PatternMatching.matching_problem
                                  | PatternMatching.ABKGoal ->
                                    ((let uu___2431450 =
                                          (let uu___2431451 =
                                              PatternMatching.pattern_of_term typ
                                              <:
                                              FStar.Tactics.Effect.TAC PatternMatching.Types.pattern
                                            in
                                            (FStar.Pervasives.Native.Some uu___2431451
                                              <:
                                              Prims.PURE
                                              (FStar.Pervasives.Native.option PatternMatching.Types.pattern
                                                ))
                                            <:
                                            FStar.Tactics.Effect.TAC
                                            (FStar.Pervasives.Native.option PatternMatching.Types.pattern
                                              ))
                                          <:
                                          FStar.Tactics.Effect.TAC
                                          (FStar.Pervasives.Native.option PatternMatching.Types.pattern
                                            )
                                        in
                                        (PatternMatching.Mkmatching_problem
                                            (Mkmatching_problem?.mp_vars (problem
                                                  <:
                                                  Prims.PURE PatternMatching.matching_problem))
                                            (Mkmatching_problem?.mp_hyps (problem
                                                  <:
                                                  Prims.PURE PatternMatching.matching_problem))
                                            uu___2431450
                                          <:
                                          Prims.PURE PatternMatching.matching_problem)
                                        <:
                                        FStar.Tactics.Effect.TAC PatternMatching.matching_problem)
                                      <:
                                      FStar.Tactics.Effect.TAC PatternMatching.matching_problem)
                                    <:
                                    FStar.Tactics.Effect.TAC PatternMatching.matching_problem)
                                <:
                                FStar.Tactics.Effect.TAC PatternMatching.matching_problem)
                              <:
                              PatternMatching.matching_problem)
                            <:
                            FStar.Tactics.Effect.TAC PatternMatching.matching_problem)
                          <:
                          FStar.Tactics.Effect.TAC PatternMatching.matching_problem)
                        <:
                        PatternMatching.matching_problem)
                    (PatternMatching.Mkmatching_problem [] [] (FStar.Pervasives.Native.None))
                    a))
            _
            (fun problem ->
                FStar.Tactics.Effect._dm4f_TAC_bind_elab Prims.unit
                  (PatternMatching.matching_problem * PatternMatching.abspat_continuation)
                  _
                  (reify (PatternMatching.debug "Done computing problem"))
                  _
                  (fun _ ->
                      FStar.Tactics.Effect._dm4f_TAC_bind_elab Prims.unit
                        (PatternMatching.matching_problem * PatternMatching.abspat_continuation)
                        _
                        (reify (PatternMatching.debug ("Got matching problem: " ^
                                  PatternMatching.string_of_matching_problem (PatternMatching.Mkmatching_problem
                                        (FStar.List.Tot.Base.rev (Mkmatching_problem?.mp_vars problem
                                              ))
                                        (FStar.List.Tot.Base.rev (Mkmatching_problem?.mp_hyps problem
                                              ))
                                        (Mkmatching_problem?.mp_goal problem)
                                      <:
                                      Prims.PURE PatternMatching.matching_problem))))
                        _
                        (fun _ ->
                            FStar.Tactics.Effect._dm4f_TAC_return_elab
                              (PatternMatching.matching_problem *
                                PatternMatching.abspat_continuation)
                              ((PatternMatching.Mkmatching_problem
                                    (FStar.List.Tot.Base.rev (Mkmatching_problem?.mp_vars problem))
                                    (FStar.List.Tot.Base.rev (Mkmatching_problem?.mp_hyps problem))
                                    (Mkmatching_problem?.mp_goal problem)
                                  <:
                                  Prims.PURE PatternMatching.matching_problem),
                                (FStar.List.Tot.Base.map (fun uu___2431455 ->
                                        (let binder, _, binder_kind, _ = uu___2431455 in
                                          PatternMatching.Mkabspat_argspec binder binder_kind
                                          <:
                                          PatternMatching.abspat_argspec)
                                        <:
                                        PatternMatching.abspat_argspec)
                                    a,
                                  FStar.Tactics.Builtins.__embed
                                  (fun _ h ->
                                      (FStar.Tactics.Builtins.clear h ()
                                        <:
                                        FStar.Tactics.Effect.TAC Prims.unit)
                                      <:
                                      Prims.unit)
                                  <:
                                  Prims.PURE
                                  (Prims.list PatternMatching.abspat_argspec *
                                    FStar.Reflection.Types.term))))))
            (FStar.Tactics.Types.decr_depth q))
        <:
        FStar.Tactics.Result.__result
        (PatternMatching.matching_problem * PatternMatching.abspat_continuation)
      | FStar.Tactics.Result.Failed (msg, q) -> FStar.Tactics.Result.Failed (msg, q)
    with
    | FStar.Tactics.Result.Success (a, q) ->
      (let () = FStar.Tactics.Types.tracepoint q in
        (let problem, continuation = a in
          FStar.Tactics.Effect._dm4f_TAC_bind_elab
            (uu___2431468:PatternMatching.matching_solution ->
              FStar.Tactics.Effect.Tac PatternMatching.matching_solution)
            PatternMatching.matching_solution
            _
            (reify (PatternMatching.tpair continuation))
            _
            (fun uu___2431469 ->
                reify (PatternMatching.solve_mp problem
                      [
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"; "(alien:reflection.embed_binder)";
                        "(alien:reflection.embed_binder)"
                      ]
                      (FStar.Tactics.Builtins.__embed (Prims.squash (a /\ b ==> c == b ==> c)))
                      uu___2431469))) (FStar.Tactics.Types.decr_depth q))
      <:
      FStar.Tactics.Result.__result
      (Prims.list PatternMatching.abspat_argspec * FStar.Reflection.Types.term *
        PatternMatching.matching_solution)
    | FStar.Tactics.Result.Failed (msg, q) -> FStar.Tactics.Result.Failed (msg, q)
  with
  | FStar.Tactics.Result.Success (a, q) ->
    (let () = FStar.Tactics.Types.tracepoint q in
      (let continuation, solution = a in
        FStar.Tactics.Effect._dm4f_TAC_bind_elab
          (uu___2431476:PatternMatching.matching_solution -> FStar.Tactics.Effect.Tac Prims.unit)
          Prims.unit
          _
          (reify (PatternMatching.interp_abspat_continuation Prims.unit continuation))
          _
          (fun uu___2431477 -> reify (uu___2431477 solution))) (FStar.Tactics.Types.decr_depth q))
    <:
    FStar.Tactics.Result.__result Prims.unit
  | FStar.Tactics.Result.Failed (msg, q) -> FStar.Tactics.Result.Failed (msg, q))
<:
FStar.Tactics.Result.__result Prims.unit
```

This time it comes with a warning:

```
Not an embedded string: "Got binder: " ^ ("a#2446390" <: Prims.PURE Prims.string) ^ "; type is Type"
```

Repro is at https://github.com/FStarLang/FStar/commit/8513bae5ab2702fdc25251da44cb65ca658d3362

**Comment by mtzguido** [comment]
Isn't this just #1272? Your first example works if you let-bind the normalized term.

**Comment by mtzguido** [comment]
And `PatternMatching.fst` on that commit just ran ok for me...? At least after removing some `uu___` identifier and running up to line 588 (there's a duplicate top-level afterwards)

**Comment by cpitclaudel** [comment]
Could be :) Feel free to close, in that case. I was tripped up by the different error message.

**Comment by cpitclaudel** [comment]
The one that's failing for me in that repro is line https://github.com/FStarLang/FStar/blob/8513bae5ab2702fdc25251da44cb65ca658d3362/examples/pattern-matching/PatternMatching.fst#L567

**Comment by mtzguido** [comment]
It works here, I swear! :)

But I think it might be a valid error: that ascription might be blocking reduction of primitive ops (`(^)`), will try to repro.

**Comment by cpitclaudel** [comment]
Oh wait, sorry

**Comment by cpitclaudel** [comment]
It's the *next* one that fails

**Comment by cpitclaudel** [comment]
Just rename it into `example2` or something else. I tend to miss these name collisions because I work in Emacs.

**Comment by mtzguido** [comment]
Not the ascription, but a `Tm_meta` around an argument that prevents reduction. Will inspect better and get back.

**Comment by mtzguido** [comment]
Just an update, this second error disappeared in my branch due to moving to weak reduction for tactics. I didn't expect it to make a difference except for errors, but it does matter when we have higher-order tactics (aka tacticals), since if we're normalizing `__trytac #_ (fun ps -> e)`, weak reduction will not do anything with `fun ps -> e` (a reified tactic) while strong reduction will (which seems to introduce a `Meta_monadic_lift` in that string you're concatenating, preventing reduction, correctly I believe).

So, this might happen again... but I think #1272 is the real issue, we're messing up `Meta_monadic`s and reification.

**Comment by mtzguido** [comment]
After (hopefully) having fixed #1272, the second part of this issue remains. Relabeling.

**Comment by mtzguido** [comment]
This now seems fixed in my branch, even with strong reduction, after some various normalizer/reification fixes. I'll merge it soon, closing preemptively...
